### PR TITLE
[do-not-merge] use a 5.16 debug build

### PR DESCRIPTION
**Pull request by lewurm**
hopefully helps to identify the msbuild hang with mono

Context: https://github.com/mono/mono/pull/11706 & https://github.com/xamarin/xamarin-android/pull/2357

/cc @luhenry

### CODE REVIEW

**Comment by lewurm** [comment]
stupid question: How do I make _sure_ the bot will use the system mono I specified in this PR? Because it doesn't seem to be using it:

```console
xam-mac-mini-22:~ builder$ mono --version
Mono JIT compiler version 5.16.0.106 (2018-06/341142d7656 Fri Aug 10 13:07:02 EDT 2018)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
	TLS:           normal
	SIGSEGV:       altstack
	Notification:  kqueue
	Architecture:  amd64
	Disabled:      none
	Misc:          softdebug
	Interpreter:   yes
	LLVM:          yes(3.6.0svn-mono-master/0b3cb8ac12c)
	GC:            sgen (concurrent by default)
```

Is that the way to go?
```patch
diff --git a/Configuration.props b/Configuration.props
index b0bf7c45f5..1a4be789fa 100644
--- a/Configuration.props
+++ b/Configuration.props
@@ -71,7 +71,7 @@
     <JavaInteropSourceDirectory Condition=" '$(JavaInteropSourceDirectory)' == '' ">$(MSBuildThisFileDirectory)external\Java.Interop</JavaInteropSourceDirectory>
     <LlvmSourceDirectory Condition=" '$(LlvmSourceDirectory)' == '' ">$(MSBuildThisFileDirectory)external\llvm</LlvmSourceDirectory>
     <MonoSourceDirectory>$(MSBuildThisFileDirectory)external\mono</MonoSourceDirectory>
-    <MonoRequiredMinimumVersion Condition=" '$(MonoRequiredMinimumVersion)' == '' ">5.16.0</MonoRequiredMinimumVersion>
+    <MonoRequiredMinimumVersion Condition=" '$(MonoRequiredMinimumVersion)' == '' ">5.16.0.219</MonoRequiredMinimumVersion>
     <MonoRequiredMaximumVersion Condition=" '$(MonoRequiredMaximumVersion)' == '' ">5.17.0</MonoRequiredMaximumVersion>
     <IgnoreMaxMonoVersion Condition=" '$(IgnoreMaxMonoVersion)' == '' ">True</IgnoreMaxMonoVersion>
     <MonoRequiredDarwinMinimumVersion>$(MonoRequiredMinimumVersion).0</MonoRequiredDarwinMinimumVersion>
```

**Comment by lewurm** [comment]
seems to pickup the right version now

**Comment by lewurm** [comment]
Got a hang here https://jenkins.mono-project.com/job/xamarin-android-pr-builder-debug/16/console with a debug build (that is, `CFLAGS='-O0 -g'`). Alas I missed the window to attach with a debugger.

Since we are certain now that we can reproduce the hang with a debug build, I'm gonna add some more logging and update to the new package.

**Comment by lewurm** [comment]
got a hang again, and with the additional debugging now, we know what process msbuild is waiting for, it is `nunit-agent.exe` (pid 67177):
```
/Library/Frameworks/Mono.framework/Versions/Current/Commands/mono --runtime=v4.0.30319 /Users/builder/jenkins/workspace/xamarin-android-pr-builder-debug/xamarin-android/external/Java.Interop/packages/NUnit.ConsoleRunner.3.7.0/tools/nunit-agent.exe 8c463d86-3ec5-4f3b-a812-4f653ccd2266 tcp://127.0.0.1:56589/TestAgency --pid=67176
```
`bt all` of 67177:
https://gist.github.com/lewurm/e466e619531e3242a9e1a46e88721815
unfortunately it is _without_ debugging symbols for unknown reasons? 🙁 and also `lldb` refuses to let me call `mono_pmip`? Anyway, it's _weird_ that `Thread 1` is waiting for the loader lock (BTW, I didn't figure out how to get the thread owner out of the mutex; how do you do that on macOS? it looks like Apple changed `pthread_mutex_t` to an opaque definition 😠 ).

Process 67176 does _not_ run anymore on that machine.

That is what we get on stdout/stderr:
```
09:16:21   Runtime Environment
09:16:21   XXX-MONO: process_create: Exec prog [/Library/Frameworks/Mono.framework/Versions/Current/Commands/mono] args [--runtime=v4.0.30319 "/Users/builder/jenkins/workspace/xamarin-android-pr-builder-debug/xamarin-android/external/Java.Interop/packages/NUnit.ConsoleRunner.3.7.0/tools/nunit-agent.exe" 8c463d86-3ec5-4f3b-a812-4f653ccd2266 tcp://127.0.0.1:56589/TestAgency --pid=67176]
09:16:21      OS Version: MacOSX 16.7.0.0 
09:16:21     CLR Version: 4.0.30319.42000
09:16:21   
09:16:21   Test Files
09:16:21       /Users/builder/jenkins/workspace/xamarin-android-pr-builder-debug/xamarin-android/external/Java.Interop/build-tools/scripts/../../bin/TestDebug/Java.Interop-PerformanceTests.dll
09:16:21   
09:16:21   XXX-MONO: process_create: returning handle 0x7fc1ac81b518 for pid 67177
09:16:22   WARNING in native method: JNI call made without checking exceptions when required to from CallStaticObjectMethod
09:21:54   Stacktrace:
09:21:54   
09:21:54     at <unknown> <0xffffffff>
09:21:54     at System.Threading.ThreadPoolGlobals..cctor () [0x0001b] in <63aaafb4c0f64c7290fef1cfdd40a9df>:0
09:21:54     at (wrapper runtime-invoke) object.runtime_invoke_void (object,intptr,intptr,intptr) [0x0001e] in <63aaafb4c0f64c7290fef1cfdd40a9df>:0
09:21:54     at <unknown> <0xffffffff>
09:21:54     at System.Threading.ThreadPool.QueueUserWorkItem (System.Threading.WaitCallback) [0x00002] in <63aaafb4c0f64c7290fef1cfdd40a9df>:0
09:21:54     at NUnit.Engine.Agents.RemoteTestAgent.Stop () [0x0000f] in <1e8ad6af4c6f4686ad7e5f9e67020b3b>:0
09:21:54     at (wrapper runtime-invoke) object.runtime_invoke_void__this__ (object,intptr,intptr,intptr) [0x0004d] in <63aaafb4c0f64c7290fef1cfdd40a9df>:0
09:21:54     at <unknown> <0xffffffff>
09:21:54     at (wrapper managed-to-native) System.Runtime.Remoting.RemotingServices.InternalExecute (System.Reflection.MethodBase,object,object[],object[]&) [0x00000] in <63aaafb4c0f64c7290fef1cfdd40a9df>:0
09:21:54     at System.Runtime.Remoting.RemotingServices.InternalExecuteMessage (System.MarshalByRefObject,System.Runtime.Remoting.Messaging.IMethodCallMessage) [0x000b5] in <63aaafb4c0f64c7290fef1cfdd40a9df>:0
09:21:54     at System.Runtime.Remoting.Messaging.StackBuilderSink.SyncProcessMessage (System.Runtime.Remoting.Messaging.IMessage) [0x0001c] in <63aaafb4c0f64c7290fef1cfdd40a9df>:0
09:21:54     at System.Runtime.Remoting.Messaging.ServerObjectTerminatorSink.SyncProcessMessage (System.Runtime.Remoting.Messaging.IMessage) [0x00016] in <63aaafb4c0f64c7290fef1cfdd40a9df>:0
09:21:54     at System.Runtime.Remoting.Lifetime.LeaseSink.SyncProcessMessage (System.Runtime.Remoting.Messaging.IMessage) [0x00007] in <63aaafb4c0f64c7290fef1cfdd40a9df>:0
09:21:54     at System.Runtime.Remoting.ClientActivatedIdentity.SyncObjectProcessMessage (System.Runtime.Remoting.Messaging.IMessage) [0x00035] in <63aaafb4c0f64c7290fef1cfdd40a9df>:0
09:21:54     at System.Runtime.Remoting.Messaging.ServerContextTerminatorSink.SyncProcessMessage (System.Runtime.Remoting.Messaging.IMessage) [0x0001a] in <63aaafb4c0f64c7290fef1cfdd40a9df>:0
09:21:54     at System.Runtime.Remoting.Contexts.CrossContextChannel.SyncProcessMessage (System.Runtime.Remoting.Messaging.IMessage) [0x0004a] in <63aaafb4c0f64c7290fef1cfdd40a9df>:0
09:21:54     at System.Runtime.Remoting.Channels.ChannelServices.SyncDispatchMessage (System.Runtime.Remoting.Messaging.IMessage) [0x00012] in <63aaafb4c0f64c7290fef1cfdd40a9df>:0
09:21:54     at System.Runtime.Remoting.Channels.ChannelServices.DispatchMessage (System.Runtime.Remoting.Channels.IServerChannelSinkStack,System.Runtime.Remoting.Messaging.IMessage,System.Runtime.Remoting.Messaging.IMessage&) [0x0000e] in <63aaafb4c0f64c7290fef1cfdd40a9df>:0
09:21:54     at System.Runtime.Remoting.Channels.ServerDispatchSink.ProcessMessage (System.Runtime.Remoting.Channels.IServerChannelSinkStack,System.Runtime.Remoting.Messaging.IMessage,System.Runtime.Remoting.Channels.ITransportHeaders,System.IO.Stream,System.Runtime.Remoting.Messaging.IMessage&,System.Runtime.Remoting.Channels.ITransportHeaders&,System.IO.Stream&) [0x00008] in <63aaafb4c0f64c7290fef1cfdd40a9df>:0
09:21:54     at System.Runtime.Remoting.Channels.BinaryServerFormatterSink.ProcessMessage (System.Runtime.Remoting.Channels.IServerChannelSinkStack,System.Runtime.Remoting.Messaging.IMessage,System.Runtime.Remoting.Channels.ITransportHeaders,System.IO.Stream,System.Runtime.Remoting.Messaging.IMessage&,System.Runtime.Remoting.Channels.ITransportHeaders&,System.IO.Stream&) [0x000b8] in <7df9d33badaa4058b34a26fb21965f57>:0
09:21:54     at NUnit.Engine.Internal.TcpChannelUtils/ObservableServerChannelSinkProvider/ObservableServerChannelSink.ProcessMessage (System.Runtime.Remoting.Channels.IServerChannelSinkStack,System.Runtime.Remoting.Messaging.IMessage,System.Runtime.Remoting.Channels.ITransportHeaders,System.IO.Stream,System.Runtime.Remoting.Messaging.IMessage&,System.Runtime.Remoting.Channels.ITransportHeaders&,System.IO.Stream&) [0x0000d] in <1e8ad6af4c6f4686ad7e5f9e67020b3b>:0
09:21:54     at System.Runtime.Remoting.Channels.Tcp.TcpServerTransportSink.InternalProcessMessage (System.Runtime.Remoting.Channels.Tcp.ClientConnection,System.IO.Stream) [0x0006d] in <7df9d33badaa4058b34a26fb21965f57>:0
09:21:54     at System.Runtime.Remoting.Channels.Tcp.ClientConnection.ProcessMessages () [0x00042] in <7df9d33badaa4058b34a26fb21965f57>:0
09:21:54     at System.Runtime.Remoting.Channels.RemotingThreadPool.PoolThread () [0x000cb] in <7df9d33badaa4058b34a26fb21965f57>:0
09:21:54     at System.Threading.ThreadHelper.ThreadStart_Context (object) [0x00014] in <63aaafb4c0f64c7290fef1cfdd40a9df>:0
09:21:54     at System.Threading.ExecutionContext.RunInternal (System.Threading.ExecutionContext,System.Threading.ContextCallback,object,bool) [0x00071] in <63aaafb4c0f64c7290fef1cfdd40a9df>:0
09:21:54     at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext,System.Threading.ContextCallback,object,bool) [0x00000] in <63aaafb4c0f64c7290fef1cfdd40a9df>:0
09:21:54     at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext,System.Threading.ContextCallback,object) [0x0002b] in <63aaafb4c0f64c7290fef1cfdd40a9df>:0
09:21:54     at System.Threading.ThreadHelper.ThreadStart () [0x00008] in <63aaafb4c0f64c7290fef1cfdd40a9df>:0
09:21:54     at (wrapper runtime-invoke) object.runtime_invoke_void__this__ (object,intptr,intptr,intptr) [0x0004d] in <63aaafb4c0f64c7290fef1cfdd40a9df>:0
09:21:54   System.Runtime.Remoting.RemotingException: Tcp transport error. ---> System.Runtime.Remoting.RemotingException: Connection closed
09:21:54     at System.Runtime.Remoting.Channels.Tcp.TcpMessageIO.StreamRead (System.IO.Stream networkStream, System.Byte[] buffer, System.Int32 count) [0x00024] in /Users/builder/jenkins/workspace/build-package-osx-mono-pullrequest/pr/external/bockbuild/builds/mono-x64/mcs/class/System.Runtime.Remoting/System.Runtime.Remoting.Channels.Tcp/TcpMessageIO.cs:95 
09:21:54     at System.Runtime.Remoting.Channels.Tcp.TcpMessageIO.ReceiveMessageStatus (System.IO.Stream networkStream, System.Byte[] buffer) [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono-pullrequest/pr/external/bockbuild/builds/mono-x64/mcs/class/System.Runtime.Remoting/System.Runtime.Remoting.Channels.Tcp/TcpMessageIO.cs:54 
09:21:54      --- End of inner exception stack trace ---
09:21:54   
09:21:54   
09:21:54   Server stack trace: 
09:21:54     at System.Runtime.Remoting.Channels.Tcp.TcpMessageIO.ReceiveMessageStatus (System.IO.Stream networkStream, System.Byte[] buffer) [0x0000c] in /Users/builder/jenkins/workspace/build-package-osx-mono-pullrequest/pr/external/bockbuild/builds/mono-x64/mcs/class/System.Runtime.Remoting/System.Runtime.Remoting.Channels.Tcp/TcpMessageIO.cs:56 
09:21:54     at System.Runtime.Remoting.Channels.Tcp.TcpClientTransportSink.ProcessMessage (System.Runtime.Remoting.Messaging.IMessage msg, System.Runtime.Remoting.Channels.ITransportHeaders requestHeaders, System.IO.Stream requestStream, System.Runtime.Remoting.Channels.ITransportHeaders& responseHeaders, System.IO.Stream& responseStream) [0x00052] in /Users/builder/jenkins/workspace/build-package-osx-mono-pullrequest/pr/external/bockbuild/builds/mono-x64/mcs/class/System.Runtime.Remoting/System.Runtime.Remoting.Channels.Tcp/TcpClientTransportSink.cs:188 
09:21:54     at System.Runtime.Remoting.Channels.BinaryClientFormatterSink.SyncProcessMessage (System.Runtime.Remoting.Messaging.IMessage msg) [0x00066] in /Users/builder/jenkins/workspace/build-package-osx-mono-pullrequest/pr/external/bockbuild/builds/mono-x64/mcs/class/System.Runtime.Remoting/System.Runtime.Remoting.Channels/BinaryClientFormatterSink.cs:157 
09:21:54   
09:21:54   Exception rethrown at [0]: 
09:21:54     at (wrapper managed-to-native) System.Object.__icall_wrapper_mono_remoting_wrapper(intptr,intptr)
09:21:54     at (wrapper remoting-invoke) NUnit.Engine.Agents.RemoteTestAgent.Stop()
09:21:54     at NUnit.Engine.Runners.ProcessRunner.Dispose (System.Boolean disposing) [0x00086] in <1e8ad6af4c6f4686ad7e5f9e67020b3b>:0 
09:21:54     at NUnit.Engine.Runners.AbstractTestRunner.Dispose () [0x00000] in <1e8ad6af4c6f4686ad7e5f9e67020b3b>:0 
09:21:54     at NUnit.Engine.Runners.MasterTestRunner.Dispose (System.Boolean disposing) [0x00013] in <1e8ad6af4c6f4686ad7e5f9e67020b3b>:0 
09:21:54     at NUnit.Engine.Runners.MasterTestRunner.Dispose () [0x00000] in <1e8ad6af4c6f4686ad7e5f9e67020b3b>:0 
09:21:54     at NUnit.ConsoleRunner.ConsoleRunner.RunTests (NUnit.Engine.TestPackage package, NUnit.Engine.TestFilter filter) [0x0010e] in <5d13e9f4d03e4da1b4779cb1d61b9b3d>:0 
09:21:54     at NUnit.ConsoleRunner.ConsoleRunner.Execute () [0x000b6] in <5d13e9f4d03e4da1b4779cb1d61b9b3d>:0 
09:21:54     at NUnit.ConsoleRunner.Program.Main (System.String[] args) [0x001bf] in <5d13e9f4d03e4da1b4779cb1d61b9b3d>:0 
```

So in conclusion: The msbuild process is waiting for process 67177 which looks fine to me. 67177 itself is weird; the wait on the loader lock is suspicious. @jonpryor does the managed stack trace tell you _anything_? Thread 5 is waiting in `ves_icall_System_Net_Sockets_Socket_Accept_internal`... which is weird as the managed stack trace indicates `Tcp transport error`?

**Comment by rolfbjarne** [comment]
> BTW, I didn't figure out how to get the thread owner out of the mutex; how do you do that on macOS? it looks like Apple changed `pthread_mutex_t` to an opaque definition 😠

This is not easy.

1. Figure out the libpthread version you have on your system:

```
$ otool -L /usr/lib/system/libsystem_pthread.dylib | head -2
/usr/lib/system/libsystem_pthread.dylib:
	/usr/lib/system/libsystem_pthread.dylib (compatibility version 1.0.0, current version 301.50.1)
```

2. Find the source code for that version [here](https://opensource.apple.com/source/libpthread).
3. Find the [header](https://opensource.apple.com/source/libpthread/libpthread-301.50.1/src/internal.h.auto.html) for the `_pthread_mutex` struct.
4. Examine your `pthread_mutex_t`'s memory and match it with the `_pthread_mutex` struct's layout to determine where the thread id of the thread that has the mutex locked is.

**Comment by jonpryor** [comment]
@lewurm asked:

> @jonpryor does the managed stack trace tell you anything?

Not really.

(Reminder?) The NUnit architecture *by default* involves a dedicated "test-runner" process; when you run `mono nunit3-console.exe`, `nunit3-console.exe` starts a sub-process, and communicates with that process via Remoting.  This is so that if the unit test fails *horribly* (SIGSEGV, exit), test results can still be captured.

For example, consider this unit test fixture:

```csharp
using System;
using System.Diagnostics;
using System.Runtime.InteropServices;

using NUnit.Framework;

namespace MyTests {

	[TestFixture]
	class DieDieDie {
		[DllImport ("c")]
		static extern void _exit (int value);

		[DllImport ("c")]
		static extern int printf (string format, string value);
		
		[Test]
		public void Foo ()
		{
			var id = Process.GetCurrentProcess ().Id;
			printf ("%s", $"# jonp: Test Runner PID: {id}\n");
			_exit (42);
		}
	}
}
```

When we compile and run `nunit3-console.exe`:

```bash
$ mono packages/NUnit.ConsoleRunner.3.7.0/tools/nunit3-console.exe Tests.dll &
[1] 41499

$ echo $!
41499

$ fg
NUnit Console Runner 3.7.0 
...
# jonp: Test Runner PID: 41500

Errors, Failures and Warnings

1) Error : 
Tcp transport error.


Server stack trace: 
  at System.Runtime.Remoting.Channels.Tcp.TcpMessageIO.ReceiveMessageStatus (System.IO.Stream networkStream, System.Byte[] buffer) [0x00017] in <56ed11c124be486e9d013a2a58f35569>:0 
...
```

The in initial `nunit3-console.exe` is pid 41499, while the unit test was executed in pid 41500.

This is what you're observing in your [previous comment](
https://github.com/xamarin/xamarin-android/pull/2440#issuecomment-441205872):

```
...
09:21:54     at (wrapper managed-to-native) System.Runtime.Remoting.RemotingServices.InternalExecute (System.Reflection.MethodBase,object,object[],object[]&) [0x00000] in <63aaafb4c0f64c7290fef1cfdd40a9df>:0
09:21:54     at System.Runtime.Remoting.RemotingServices.InternalExecuteMessage (System.MarshalByRefObject,System.Runtime.Remoting.Messaging.IMethodCallMessage) [0x000b5] in <63aaafb4c0f64c7290fef1cfdd40a9df>:0
09:21:54     at System.Runtime.Remoting.Messaging.StackBuilderSink.SyncProcessMessage (System.Runtime.Remoting.Messaging.IMessage) [0x0001c] in <63aaafb4c0f64c7290fef1cfdd40a9df>:0
...
```

That stack trace is *presumably* the "calling" `nunit3-console.exe` execution waiting on the sub-process to either send it a message or otherwise exit.  (Presumably.)

Assuming that your pid 67177 is the `nunit3-console.exe` process, I don't know why it would be waiting on the loader lock.

**Comment by lewurm** [comment]
> unfortunately it is without debugging symbols for unknown reasons? 🙁

still no idea why, but at least I found a workaround:
```
(lldb) add-dsym /Library/Frameworks/Mono.framework/Commands/mono-sgen.dSYM/Contents/Resources/DWARF/mono-sgen64
```

**Comment by lewurm** [comment]
looking at https://jenkins.mono-project.com/job/xamarin-android-pr-builder-release/129/console

some observations:
* 
```
XXX-MONO: process_create: Exec prog [/Library/Frameworks/Mono.framework/Versions/Current/Commands/mono] args [--runtime=v4.0.30319 "/Users/builder/jenkins/workspace/xamarin-android-pr-builder-release/xamarin-android/external/Java.Interop/packages/NUnit.ConsoleRunner.3.7.0/tools/nunit-agent.exe" 6303eefd-4272-4b6e-8335-d4b762f3cd21 tcp://127.0.0.1:60147/TestAgency --pid=49322
XXX-MONO: process_create: returning handle 0x7fb88f818f18 for pid 49323 from 49322
```
Process 49322 doesn't exist anymore
* Process 49323 is stuck on the loader lock trying to compile `System.Diagnostics.Process:RaiseOnExited ()`
* ~~I can't use `mono_pmip` because the loader lock is taken~~
* this thread _was_ the owner of the loader lock
```
(lldb) p/x *(((guint32*) &loader_mutex)+6)
(guint32) $16 = 0x000f390c
```
* list of threads alive:
```
(lldb) thread list
Process 49323 stopped
* thread #1: tid = 0xf3902, 0x00007fffc9bc0c22 libsystem_kernel.dylib`__psynch_mutexwait + 10, name = 'tid_307', queue = 'com.apple.main-thread', stop reason = signal SIGSTOP
  thread #2: tid = 0xf3903, 0x00007fffc9bc0bf2 libsystem_kernel.dylib`__psynch_cvwait + 10, name = 'SGen worker'
  thread #3: tid = 0xf3904, 0x00007fffc9bb9386 libsystem_kernel.dylib`semaphore_wait_trap + 10, name = 'Finalizer'
  thread #4: tid = 0xf3905, 0x00007fffc9bc144e libsystem_kernel.dylib`__workq_kernreturn + 10
  thread #5: tid = 0xf3908, 0x00007fffc9bc00c2 libsystem_kernel.dylib`__accept + 10, name = 'tid_210b'
  thread #6: tid = 0xf3909, 0x00007fffc9bc0bf2 libsystem_kernel.dylib`__psynch_cvwait + 10, name = 'tid_2e03'
  thread #7: tid = 0xf3913, 0x00007fffc9bc0bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
  thread #8: tid = 0xf3914, 0x00007fffc9bc0bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
  thread #9: tid = 0xf3915, 0x00007fffc9bc0bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
  thread #10: tid = 0xf3916, 0x00007fffc9bc0bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
  thread #11: tid = 0xf3917, 0x00007fffc9bc0bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
  thread #12: tid = 0xf3918, 0x00007fffc9bc0bf2 libsystem_kernel.dylib`__psynch_cvwait + 10, name = 'Java: Reference Handler'
  thread #13: tid = 0xf3919, 0x00007fffc9bc0bf2 libsystem_kernel.dylib`__psynch_cvwait + 10, name = 'Java: Finalizer'
  thread #14: tid = 0xf391a, 0x00007fffc9bb9386 libsystem_kernel.dylib`semaphore_wait_trap + 10, name = 'Java: Signal Dispatcher'
  thread #15: tid = 0xf391b, 0x00007fffc9bc0bf2 libsystem_kernel.dylib`__psynch_cvwait + 10, name = 'Java: C2 CompilerThread0'
  thread #16: tid = 0xf391c, 0x00007fffc9bc0bf2 libsystem_kernel.dylib`__psynch_cvwait + 10, name = 'Java: C2 CompilerThread1'
  thread #17: tid = 0xf391d, 0x00007fffc9bc0bf2 libsystem_kernel.dylib`__psynch_cvwait + 10, name = 'Java: C1 CompilerThread2'
  thread #18: tid = 0xf391e, 0x00007fffc9bc0bf2 libsystem_kernel.dylib`__psynch_cvwait + 10, name = 'Java: Service Thread'
  thread #19: tid = 0xf391f, 0x00007fffc9bc0bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
```
* tried to "force" unlock the loader mutex, but no success.
*
```
XXX-MONO: process_create: Exec prog [/bin/sh] args [ -c "export LANG=en_US.UTF-8; export LC_ALL=en_US.UTF-8; . /var/folders/qw/n987bpf502dgctv9n_v7s27w0000w3/T/tmpa62933807d37403e95ba884f726a7d41.exec.cmd"]
XXX-MONO: process_create: returning handle 0x7f9b28804780 for pid 49321 from 29372
```
where 29372 is msbuild. 49321 is `sh`, spawning `nunit3-console.exe` (presumably 49322):
```console
$ cat /var/folders/qw/n987bpf502dgctv9n_v7s27w0000w3/T/tmpa62933807d37403e95ba884f726a7d41.exec.cmd
#!/bin/sh
mono --debug packages/NUnit.ConsoleRunner.3.7.0/tools/nunit3-console.exe  /Users/builder/jenkins/workspace/xamarin-android-pr-builder-release/xamarin-android/external/Java.Interop/build-tools/scripts/../../bin/TestRelease/Java.Interop-PerformanceTests.dll  --result="TestResult-Java.Interop-PerformanceTests.xml;format=nunit2" --output="bin/TestRelease/TestOutput-Java.Interop-PerformanceTests.txt"
```
* msbuild `29372` is waiting for some process to finish (I think)
* I would expect msbuild to wait for `49321`, but this process doesn't exist anymore either
* attaching with `lldb` to `49323` and doing `call (void) exit(0)` results into msbuild to continue. somehow they have a relationship (child-child process?) which I don't fully understand, but the conclusion is: the loader lock hang is the issue we have to figure out.
